packages/core/src/messagingProtocols/mls/E2EIdentityService/Steps/DpopChallenge/DpopChallenge.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Wire
+ * Copyright (C) 2023 Wire Swiss GmbH
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import {DoWireDpopChallengeParams, GetClientAccessTokenParams, GetClientNonceParams} from './DpopChallenge.types';
+
+const getClientNonce = async ({apiClient, clientId}: GetClientNonceParams) => {
+  try {
+    const nonce = await apiClient.api.client.getNonce(clientId);
+    if (nonce) {
+      return nonce;
+    }
+    throw new Error('No client-nonce received');
+  } catch (e) {
+    throw new Error(`Error while trying to receive a nonce with cause: ${e}`);
+  }
+};
+
+const getClientAccessToken = async ({
+  apiClient,
+  clientNonce,
+  identity,
+  clientId,
+  expirySecs,
+}: GetClientAccessTokenParams) => {
+  const dpopToken = identity.createDpopToken(expirySecs, clientNonce);
+
+  // Remove this when the server is ready to accept the token
```

Comment:
**question**: I assume the server is not ready yet? 😅

Reply:
Yes, we still have a problem: the server is too slow for the frontend. We have the timeout in to ensure that the server is ready to be able to process the token.